src/kakarot/instructions/system_operations.cairo

```diff
@@ -152,37 +137,182 @@ namespace SystemOperations {
+    func prepare_args{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(ctx: model.ExecutionContext*, with_value: felt) -> (
+        ctx: model.ExecutionContext*, call_args: CallArgs
+    ) {
         alloc_locals;
-        let (stack, popped) = Stack.pop_n(self=ctx.stack, n=7);
+        let (stack, popped) = Stack.pop_n(self=ctx.stack, n=6 + with_value);
         let ctx = ExecutionContext.update_stack(ctx, stack);
 
         let gas = 2 ** 128 * popped[0].high + popped[0].low;
         let address = 2 ** 128 * popped[1].high + popped[1].low;
-        let value = 2 ** 128 * popped[2].high + popped[2].low;
-        let args_offset = 2 ** 128 * popped[3].high + popped[3].low;
-        let args_size = 2 ** 128 * popped[4].high + popped[4].low;
-        // TODO: use these return data inputs when RETURN is fixed: RETURN should fill the ctx.return_data and set return_data_len
-        // TODO: but not alloc a new segment
-        let ret_offset = 2 ** 128 * popped[5].high + popped[5].low;
-        let ret_size = 2 ** 128 * popped[6].high + popped[6].low;
+        let value = (2 ** 128 * popped[2].high + popped[2].low) * with_value;
+        let args_offset = 2 ** 128 * popped[2 + with_value].high + popped[2 + with_value].low;
+        let args_size = 2 ** 128 * popped[3 + with_value].high + popped[3 + with_value].low;
+        let ret_offset = 2 ** 128 * popped[4 + with_value].high + popped[4 + with_value].low;
+        let ret_size = 2 ** 128 * popped[5 + with_value].high + popped[5 + with_value].low;
+
+        // Note: We store the offset here because we can't pre-allocate a memory segment in cairo
+        // During teardown we update the memory using this offset
+        let return_data: felt* = alloc();
+        assert [return_data] = ret_offset;
 
         // Load calldata from Memory
         let (calldata: felt*) = alloc();
         let memory = Memory.load_n(
             self=ctx.memory, element_len=args_size, element=calldata, offset=args_offset
         );
-        let ctx = ExecutionContext.update_memory(ctx, memory);
 
-        // Prepare execution sub context
-        // TODO: use gas_limit when init_at_address is updated
-        let sub_ctx: model.ExecutionContext* = ExecutionContext.init_at_address(
+        let call_args = CallArgs(
+            gas=gas,
             address=address,
-            calldata_len=args_size,
-            calldata=calldata,
             value=value,
-            parent_context=ctx,
+            args_size=args_size,
+            calldata=calldata,
+            ret_size=ret_size,
+            return_data=return_data + 1,
         );
 
-        return sub_ctx;
+        let ctx = ExecutionContext.update_memory(ctx, memory);
+
+        return (ctx, call_args);
+    }
+
+    // @notice The teardown of CALLs is made in the run loop of the EVMInstructions.
+    // @return The pointer to the context
+    func teardown_call{
```

Comment:
Could this function name/functionality be more explicit?
I found: 
> In the context of Python programming and the pytest framework, a teardown is a specific type of code that is run after a test has been completed. This code is used to clean up any resources that were allocated or modified during the test, such as closing open files or connections, or rolling back changes to a database. Teardown code is typically placed in a tearDown() method in a test class, and it is automatically run by the pytest framework after each test is completed. This ensures that the tests are self-contained and do not affect each other or the overall state of the system.

Is this what you meant?

Reply:
well I named this influenced by the pytest fixture actually but I don't really know how to name this. Open to suggestion; `finalize_call`?

Reply:
I'm not sure, what does the function do? Or what is it's consequence?

Reply:
Well, it updates the calling context's stack (with success/failing) and memory (as the memory is immutable, the sub context cannot update it directly).

All the CALLs are almost the same, this is the factorization of the end of these ops (as `prepare_args` is the factorization of the beginning of these ops).